### ISSUE TITLE
Parallelize metadata processing and reduce metadata footprint

### ISSUE THREAD

**Issue by osopardo1** [Qbeast-io/qbeast-spark#343]
We noticed that we were not managing the metadata most efficiently:

1. The `_delta_log/` was read multiple times to rebuild the same data structure
2. We started to have issues when the metadata was too big
3. We were forcing a lot of computation in the Driver by executing `.toLocalIterator` in `QueryExecution`

Moving much of the computation to the Dataset API instead of using `core` classes (independent of Spark) should help reduce the number of tasks and improve resource usage.

**Comment by osopardo1** [comment]
This issue is to keep track of the changes on PR #335, which has grown more than expected. 

We measured the performance improvement by running a Sampling query over a [telemetry dataset](https://github.com/microsoft/TelemetryLogsGeneratorAndBenchmark). 
```
spark.sql("SELECT count(*) FROM logs_tpc TABLESAMPLE( 0.1 PERCENT)").show()
```

With a single node: 
1. The previous version (v0.6.0) took more than 20 seconds, and it produced all these tasks
![image](https://github.com/Qbeast-io/qbeast-spark/assets/1544874/bf696af1-57fd-42e3-bb56-6f1b9b62bd0e)
6. Then I tested this PR, and it took ~6 seconds, generating only these tasks.
![image](https://github.com/Qbeast-io/qbeast-spark/assets/1544874/69422bd2-8d9b-4d3c-b26a-38f4cf59c994)
